bpo-36817: Add f-string debugging using '='.

ericvsmith opened the pull request:
<!-- issue-number: [bpo-36817](https://bugs.python.org/issue36817) -->
https://bugs.python.org/issue36817
<!-- /issue-number -->

ericvsmith committed:
WIP: Mostly works, except repr isn't called (so only string expressions currently work).

ericvsmith committed:
WIP: Change !x to !d, to match an earlier discussion on python-ideas.

ericvsmith committed:
Call repr on the resulting expression. Change expr_source to expr_text.

ericvsmith committed:
Add simple tests.

ericvsmith committed:
Remove FVC_DEBUG, which I thought I was going to use for this feature, but it ended up not being needed.

ericvsmith committed:
Fix some comments and formatting.

ericvsmith committed:
Added blurb.

ericvsmith committed:
Use quotes around the string result in the blurb text.

ericvsmith committed:
Call out the conversion specifiers in monospace.

ericvsmith committed:
Note that the repr of the expression is used.

ericvsmith committed:
Fix conversion character.

ericvsmith committed:
Added Whats New entry.

ericvsmith committed:
Don't allow format_spec with !d.

ericvsmith committed:
Test that we don't allow a format spec.

ericvsmith committed:
Use repr(expr) if no format spec, use format(expr, spec) if one is.

ericvsmith committed:
Pre-append the '=' to the end of expr_text at compile time, instead of adding it at runtime.

ericvsmith committed:
Add a coding comment, a note about why I'm using a unicode identifier, and a test with a unicode literal.

ericvsmith committed:
Added tests for nested f-strings and for pre- and post-text.

ericvsmith committed:
Add a test for newlines in expressions.

ericvsmith committed:
Clarify a comment.

ericvsmith committed:
Change from di to =.

ericvsmith committed:
Fix a comment.

ericvsmith committed:
Fix comment.

ericvsmith committed:
Fix comments.

ericvsmith committed:
Skip all whitespace after the =, not just spaces. Handle unparsing.

ericvsmith committed:
Hack: Modify tests for f-string annotations. I'm not sure these are even valid tests any more.

ericvsmith committed:
Add missing FVC_ASCII.

ericvsmith committed:
Move the mapping from "no converion specified" to the exact conversion
function in to the compile phase.

ericvsmith committed:
Make ast_unparse understand =.

ericvsmith committed:
Add a test for !s.

ericvsmith committed:
Improve comments.

tirkarthi commented on Doc/whatsnew/3.8.rst:
```diff
@@ -148,6 +148,20 @@ extensions compiled in release mode and for C extensions compiled with the
+f-strings now support !d for quick and dirty  debugging
+-------------------------------------------------------
+
+Add ``!d`` conversion specifier to f-strings. ``f'{expr!d}'`` expands
```
Since this is an alternative to PR that implements `!d` I am just adding a note to update this in case this PR is merged to use `=` in 3.8 .

ericvsmith committed:
Update blurb for = syntax.

ericvsmith committed:
Fix whatsnew.

ericvsmith committed:
Fix comments about FVC_ values.

serhiy-storchaka commented on Parser/Python.asdl:
```diff
@@ -76,7 +76,7 @@ module Python
          | Compare(expr left, cmpop* ops, expr* comparators)
          | Call(expr func, expr* args, keyword* keywords)
-         | FormattedValue(expr value, int? conversion, expr? format_spec)
+         | FormattedValue(expr value, int? conversion, expr? format_spec, string? expr_text)
```
It is not needed and I prefer to not add redundant attributes. Instead you can add the Constant item just before FormattedValue into JoinedStr.

ericvsmith replied:
It's used in ast_unparse.c to know if the = needs to be added back in. It's not strictly necessary, of course. It depends how you feel about the fidelity required in string annotations.

serhiy-storchaka replied:
I think it if fine if `f'{foo=}'` be transformed into `f'foo={foo!r}'`. Think about this as a kind of macros.

ericvsmith replied:
I discussed this very point with Łukasz yesterday. I think his point is that he'd like to not lose the information in the unparser. Admittedly it's not perfect, and it probably can't ever be without a ton of work and additions to the AST.

serhiy-storchaka commented on Lib/test/test_future.py:
```diff
@@ -250,11 +250,21 @@ def test_annotations(self):
-        eq("f'space between opening braces: { {a for a in (1, 2, 3)}}'")
-        eq("f'{(lambda x: x)}'")
-        eq("f'{(None if a else lambda x: x)}'")
+        eq("f'some f-string with {a!f} {few()!f:.2f} {formatted.values!r}'")
```
Why these tests were changed? Can you preserve the existing tests and add new tests?

ericvsmith replied:
I'll double check. These might (or might not) be left over from the !d version. The ast unparser code has been through a number of revisions.

ericvsmith replied:
They were changed because at one point the AST didn't have enough information to know whether !f had been used in the original source, so the unparser always added !f (since it was the default). This is another case where the exact fidelity of the annotations unparser is an issue.

But in the current implementation, the !f isn't added unless it's explicitly provided, so I can undo these changes.

ericvsmith replied:
I restored these. Thanks for catching it.

serhiy-storchaka commented on Lib/test/test_fstring.py:
```diff
@@ -1049,6 +1064,36 @@ def test_backslash_char(self):
         self.assertEqual(eval('f"\\\n"'), '')
         self.assertEqual(eval('f"\\\r"'), '')
 
+    def test_debug_conversion(self):
```
Please add also tests for expressions containing "=": with operators "==", "!=", "<=", ">=", ":=" and string literals containing "=" and for format spec containing "=" (`f'{x:=10}'` where `x` is a number).

ericvsmith replied:
Good idea.

ericvsmith replied:
I've added these tests.

ericvsmith committed:
Restored some f-string tests, this time without !f.

ericvsmith committed:
Fix and improve a comment.

ericvsmith committed:
Use Py_ISSPACE instead of isspace.

ericvsmith committed:
Fix parsing error with >= and <=. Add test for operators containing =.

ericvsmith committed:
< and > without a trailing = don't end an expression.

ericvsmith committed:
Add a clarifying comment about a test.

ericvsmith committed:
Add tests for named parameters, to exercise the = parsing.

ericvsmith committed:
Improve the tests for named parameters.

ericvsmith committed:
Fix a typo in a comment.

ericvsmith committed:
Removed !f conversion. Switch to repr if using = and a format spec is supplied.

ericvsmith committed:
Add more tests for = and repr/format.

ericvsmith committed:
Fixed string annotations checks.

ericvsmith committed:
Removed unreachable code left over from !f.

ericvsmith committed:
Removed 'f' type conversion.

ericvsmith committed:
Fixed a spurious space.

larryhastings commented on Lib/test/test_fstring.py:
```diff
@@ -1049,6 +1064,90 @@ def test_backslash_char(self):
+
+        x = 20
+        # This isn't an assignment expression, it's 'x', with a format
+        # spec of '=10'.
```
It's fussy and optional, but as long as you're telling the reader it's not an assignment expression, you might add:

(Assignment expressions are supported in f-strings but they must be enclosed in parentheses.)

larryhastings commented on Lib/test/test_fstring.py:
```diff
@@ -1049,6 +1064,90 @@ def test_backslash_char(self):
+        # Make sure nested fstrings still work.
+        self.assertEqual(f'{f"{3.1415=:.1f}":*^20}', '*****3.1415=3.1*****')
+
+        # Make sure text before and after !d works correctly.
```
Fix mention of "!d"

ericvsmith committed:
Fixed whitespace.

larryhastings commented on Misc/NEWS.d/next/Core and Builtins/2019-05-02-11-48-08.bpo-36774.ZqbJ1J.rst:
```diff
@@ -0,0 +1,7 @@
+Add a ``=`` conversion feature to f-strings. This can precede ``!s``,
```
I think this text is a remnant of when it was "!d".  Back then it was a new "conversion"; now it's its own separate thing and is *not* a conversion.  Maybe you're better off choosing a different term here, just to avoid muddying the waters?  How about this for a first sentence:

    Add a ``=`` feature to f-strings for debugging.

larryhastings commented on Misc/NEWS.d/next/Core and Builtins/2019-05-02-11-48-08.bpo-36774.ZqbJ1J.rst:
```diff
@@ -0,0 +1,7 @@
+an equal sign, followed by the repr of the value of the expression. So
+``f'{3*9+15=}'`` would be equal to the string ``'3*9+15=42'``.  If
+``=`` is specified, the default conversion is set to ``!r``.  Added a
+``!f`` conversion to switch back to the previous format behavior, if
```
"!f" is now removed, and the default is now repr-unless-there-is-a-format-spec etc etc.

larryhastings commented on Python/ast.c:
```diff
@@ -5016,7 +5017,12 @@ fstring_find_expr(const char **str, const char *end, int raw, int recurse_lvl,
+    int conversion = -1; /* The conversion char.  Use default if not
+                            specified, or !r if using = and no format
+                            spec. */
+    int equal_conversion = 0; /* Are we using the = conversion? */
```
Not a "conversion" anymore.  Rename variable?  Change comment?  (Too fussy?)

ericvsmith replied:
I'm just going to leave this one, for lack of a better name. I'll fix the comment.

larryhastings commented on Python/ast.c:
```diff
@@ -5170,7 +5192,20 @@ fstring_find_expr(const char **str, const char *end, int raw, int recurse_lvl,
-        return -1;
+        goto error;
+
+    /* Check for = conversion. */
```
Just noting all the places you call '=' a *conversion* in case you want to use a different term.  Here's another!

larryhastings commented on Python/ast.c:
```diff
@@ -5170,7 +5192,20 @@ fstring_find_expr(const char **str, const char *end, int raw, int recurse_lvl,
+        *str += 1;
+        equal_conversion = 1;
+
+        /* Skip over ASCII whitespace.  No need to test for end of string
```
IIUC the comment is misleading.  If we have the malformed f-string

    f'{x='

then at this point there *isn't* "a } somewhere ahead".  Py_ISSPACE('\0') will return 0, so you'll exit the loop there, and you'll complain that you didn't stop on a '}'.  So the code works fine.  But not because a } is guaranteed--it's not like you've already validated the f-string or anything.

larryhastings commented on Python/ast_unparse.c:
```diff
@@ -655,6 +655,11 @@ append_formattedvalue(_PyUnicodeWriter *writer, expr_ty e, bool is_format_spec)
     Py_DECREF(temp_fv_str);
 
+    if (e->v.FormattedValue.expr_text) {
+        /* Use the = debug conversion. */
```
Another place you call it a "conversion".

larryhastings commented on Python/compile.c:
```diff
@@ -3946,8 +3946,8 @@ compiler_formatted_value(struct compiler *c, expr_ty e)
-       Convert the conversion char to 2 bits:
-       None: 000  0x0  FVC_NONE
+       Convert the conversion char to 3 bits:
+       !f  : 000  0x0  FVC_NONE   The default if nothing specified.
```
"!f" has been removed, is this still relevant?

larryhastings reviewed (commented):
Minor nonsense, mostly about comments.  I didn't mark this as "Approve" just because I wanted to see what you had to say in reply; if you tell me it's all fine then I'll probably just Approve it then.

ericvsmith committed:
Address review findings.

larryhastings commented on Lib/test/test_fstring.py:
```diff
@@ -1147,6 +1148,15 @@ def __repr__(self):
+    def test_walrus(self):
+        x = 20
+        # This isn't an assignment expression, it's 'x', with a format
+        # spec of '=10'.  See test_walrus: you need to use parens.
```
def test_walrus(self):
        ...
        # See test_walrus:

larryhastings commented on Python/ast.c:
```diff
@@ -5020,7 +5020,7 @@ fstring_find_expr(const char **str, const char *end, int raw, int recurse_lvl,
                             specified, or !r if using = and no format
                             spec. */
-    int equal_conversion = 0; /* Are we using the = conversion? */
+    int equal_conversion = 0; /* Are we using the = feature? */
```
So you still call the variable "equal_conversion"?  I figured you'd rename it.  "equal_debugging" maybe?

larryhastings reviewed (changes requested):
OH WHY DO PEOPLE COMPLAIN ABOUT THIS PIDDLING LITTLE STUFF

MY LIFE

ericvsmith committed:
Minor comments and variable renaming.

gpshead committed:
bpo-36816: Update the self-signed.pythontest.net cert (GH-13192)



We updated the server, our testsuite must match.


https://bugs.python.org/issue36816


✈️ CLE -> DEN ✈️ #pycon2019

(cherry picked from commit 6bd81734de0b73f1431880d6a75fb71bcbc65fa1)
